docs/design/password-key-derivation.md

```diff
@@ -0,0 +1,76 @@
+# Title: Password Key Derivation
+
+## Abstract
+
+This design doc describes the way cryptographic encryption keys will be derived from user provided passwords in a way that resists offline attacks and low entropy passwords.
+
+## Background
+
+A unique encryption key exists for every path inside of a bucket, and they are designed in a way that allows deriving a key from a parent and some path component. For example, if you had the key for `sj://bucket/path1/path2` then you could derive the key for `sj://bucket/path1/path2/path3`.
+
+That leaves open the question for how a root key is derived. Some requirements on that root key derivation include
+
+1. It's based on the user's choice (for example, a password).
+2. If the user inputs the same password on different machines, the same root key should be derived.
+
+These requirements allow users to be in full control of their encryption, and don't require users safely transporting high entropy (hard to remember) secrets to bootstrap new uplinks.
+
+This design accomodates more requirements that allow for additional features:
+
+3. A root key can be derived for any path in a bucket.
+4. Buckets root keys should not depend on the name of the bucket.
+5. A table of keys for low entropy passwords should not be possible. In other words, an attacker with knowledge of the algorithm should not be able to use a dictionary of common passwords and pre-compute what keys to check in the event of a data breach.
+6. There is a default that allows deriving keys for buckets when no bucket specific secret is provided.
+
+The third requirement allows having multiple encrypted domains to exist within a single bucket, allowing delegation of encryption. The fourth requirement allows renaming buckets without having to re-encrypt all of the paths inside of a bucket. The fifth requirement enhances security if a satellite breach happens. This is usually accomplished with [salting](https://en.wikipedia.org/wiki/Salt_(cryptography)). The sixth requirement allows providing a finite amount of information to a third party that allows them access to the keys for any bucket created in the future.
+
+## Design
+
+Uplink always stores derived keys, never the input material.
+
+The satellite will add a new endpoint, `GetSaltEntropy` that uses the api key to find the project, and optionally takes a bucket name. It is important that the result is stable, even if the bucket is renamed. For example the result of `GetSaltEntropy("bucket1")` should be the same as `GetSaltEntropy("bucket2")` even after `"bucket1"` is renamed to `"bucket2"`. Note that there are still significant challenges to bucket renaming, and we don't support it right now.
+
+When deriving a root key from an api key, password, optional bucket name and optional encrypted path, the following algorithm will be used
+
+```
+saltEntropy = GetSaltEntropy(bucketName or None)
```

Comment:
What happens when bucketName is None? Is it still meaningful to ask for a salt in that case? Would this be equivalent to returning the project UUID?

Reply:
Yes, this would be some random bytes associated to the project. I'm reluctant to tie it to the UUID, but yes, it's that.

Reply:
What would one do with the project-level random bytes? It doesn't seem like it would help in creating buckets. I gather that the intent is probably to allow a way for a whole project to share one passphrase, but there doesn't seem to be any way left for bucket encryption keys to be derived from a project root encryption key. Maybe the whole-project-passphrase use case just reduces to storing the passphrase and using it for each bucket opened, as needed?

Reply:
> What would one do with the project-level random bytes? It doesn't seem like it would help in creating buckets.

Good point. I didn't think this through well enough.

> Maybe the whole-project-passphrase use case just reduces to storing the passphrase and using it for each bucket opened, as needed?

That's essentially what the first version of the proposal was: we stored the output of the passphrase run through argon2 at a high work factor, and used that as the passphrase for any default bucket using a specific bucket salt provided by the satellite.

The thing this second proposal attempted to avoid was requesting the salt unless a passphrase was being turned into a root key. We could do something like `bucketRootKey = hmac(projectRootKey, bucketName)` the but that then means we give up on renaming buckets.

---

I have three options and I don't know which is best:

## Option 1

Update this to bring back the whole-project-passphrase and describe that clients should generate a default bucket key using that whole-project-passphrase lazily and store the result in the scope.

Then we only make network requests for a salt when using a passphrase, or the first time a new bucket is accessed that uses the whole-project-passphrase.

## Option 2

Go back to the original proposal and have each bucket just contain some stable salt data that is provided to the client whenever they request path encryption information.

## Option 3

Give up on bucket renames forever and do something else entirely.

Reply:
Optimistically implemented option 2.